File: Python/Sequences/(238)ProductOfArrayExceptSelf.py
# Time: O(n)
# Space: O(1)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        answer = []
        
        answer.append(1)
        
        for i in range(1, len(nums)):
            answer.append(nums[i-1] * answer[i-1])
            
        right = 1
        
        for i in reversed(range(len(nums))):
            answer[i] = answer[i] * right
            right = right * nums[i]
            
        return answer

File: C++/Sequences/(238)ProductOfArrayExceptSelf.cpp
// Time: O(n)
// Space: O(1)
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> answer;
        
        answer.push_back(1);
        
        for (int i = 1; i < nums.size(); i++) {
            answer.push_back(nums[i-1] * answer[i-1]);
        }
        
        int right = 1;
        
        for (int i = nums.size() - 1; i >= 0; i--) {
            answer[i] = answer[i] * right;
            right *= nums[i];
        }
        
        return answer;
    }
};